File: backend/app/core/security.py
"""Password hashing and JWT token utilities.

Why bcrypt directly instead of passlib?
  passlib 1.7.x has friction with bcrypt 4.x (deprecation warnings, missing
  __about__ attribute). For a fresh project, calling bcrypt directly is
  smaller, clearer, and avoids version-pinning gymnastics.

Why PyJWT instead of python-jose?
  PyJWT is actively maintained, has a simpler API, and ships sane defaults.
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any, Literal

import bcrypt
import jwt

from app.config import settings

TokenType = Literal["access", "refresh"]


def hash_password(password: str) -> str:
    """One-way hash a plaintext password. Cost factor = bcrypt default (12)."""
    return bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode("utf-8")


def verify_password(password: str, hashed: str) -> bool:
    """Constant-time check of plaintext against stored hash."""
    try:
        return bcrypt.checkpw(password.encode("utf-8"), hashed.encode("utf-8"))
    except (ValueError, TypeError):
        return False


def _create_token(
    subject: str,
    kind: TokenType,
    expires_delta: timedelta,
) -> str:
    now = datetime.now(timezone.utc)
    payload: dict[str, Any] = {
        "sub": subject,
        "type": kind,
        "iat": int(now.timestamp()),
        "exp": int((now + expires_delta).timestamp()),
    }
    return jwt.encode(payload, settings.jwt_secret, algorithm=settings.jwt_algorithm)


def create_access_token(subject: str | int) -> str:
    return _create_token(
        str(subject),
        "access",
        timedelta(minutes=settings.access_token_minutes),
    )


def create_refresh_token(subject: str | int) -> str:
    return _create_token(
        str(subject),
        "refresh",
        timedelta(days=settings.refresh_token_days),
    )


def decode_token(token: str) -> dict[str, Any]:
    """Decode + verify a token. Raises jwt.PyJWTError on invalid/expired."""
    return jwt.decode(token, settings.jwt_secret, algorithms=[settings.jwt_algorithm])

File: backend-ts/src/lib/security.ts
/**
 * Password hashing + JWT helpers.
 *
 * Mirrors the Python app.core.security module so an access token issued by
 * either backend is accepted by the other (same secret, same algorithm,
 * same claim shape: { sub, exp, iat, type }).
 */
import bcrypt from "bcryptjs";
import jwt, { type SignOptions, type JwtPayload } from "jsonwebtoken";
import { config } from "../config.js";

const BCRYPT_ROUNDS = 12;

export async function hashPassword(plain: string): Promise<string> {
  return bcrypt.hash(plain, BCRYPT_ROUNDS);
}

export async function verifyPassword(plain: string, hashed: string): Promise<boolean> {
  return bcrypt.compare(plain, hashed);
}

type TokenType = "access" | "refresh";

interface TokenClaims extends JwtPayload {
  sub: string;
  type: TokenType;
}

function sign(userId: number, type: TokenType, ttlSeconds: number): string {
  const opts: SignOptions = {
    algorithm: config.JWT_ALGORITHM,
    expiresIn: ttlSeconds,
  };
  return jwt.sign({ sub: String(userId), type }, config.JWT_SECRET, opts);
}

export function createAccessToken(userId: number): string {
  return sign(userId, "access", config.ACCESS_TOKEN_MINUTES * 60);
}

export function createRefreshToken(userId: number): string {
  return sign(userId, "refresh", config.REFRESH_TOKEN_DAYS * 24 * 60 * 60);
}

export function decodeToken(token: string): TokenClaims {
  const decoded = jwt.verify(token, config.JWT_SECRET, {
    algorithms: [config.JWT_ALGORITHM],
  });
  if (typeof decoded === "string") throw new Error("Unexpected string payload");
  return decoded as TokenClaims;
}
